### DISCUSSION TITLE
handling JWTs with vitedge

### DISCUSSION THREAD

**Discussion by calebeaires** [frandiox/vitedge#61] [General]
After some seeking at web on how to handling JWTs on frontend clients that fits vitedge, I did not had some success on finding the best way to do it. As usually we save token on localstorage then make requests with it, but this is not so simple at server sider rendering.

There is a to-do at docs about how to "Auth utilities/guide (passing JWT in requests as cookies)". Is there someone from vitedge community to help us delivery it?

**Comment by frandiox** [comment]
@calebeaires Yep, I added it to the roadmap because I thought this could be challenging.
I haven't done this myself yet so I don't have actual code but I can probably give you some pointers.

The main problem when using JWTs is that they can't be passed automatically when making a GET request for an SSR rendered page.
However, cookies are sent automatically by the browser.
Therefore, I think you can basically store your JWT inside a cookie (ideally HTTPOnly cookie, which is safer). That way, you'll be able to read cookies in your backend and get the JWT from there.
For example, [here](https://medium.com/@ryanchenkie_40935/react-authentication-how-to-store-jwt-in-a-cookie-346519310e81)'s a blog post about storing JWT in cookies.

This approach should work generally but I haven't tried it so there might be some wild bug. Please let me know when you try it.

**Reply by calebeaires** [reply]
Thanks again. 

Could you lead me to access those cookies. I have tried this way, but no success. Here what I have tried

```js
// For a JS project, '.js' extension MUST be provided due to ESM limitations
import {ApiEndpoint} from 'vitedge'
import {staticImportTestWithExtension} from '../../utils/test1.js'
// For a TS project, '.js' extension can be optionally provided
import {staticImportTestWithoutExtension} from '../../utils/test2'
// JSON can be imported because --experimental-json-modules is provided in CLI dev
import json from '../../utils/test3.json'

export default <ApiEndpoint>{
    handler({query, request, headers}) {
        if (request.method !== 'GET') {
            throw new Error('Method not supported!')
        }

// tried to access cookies

console.log(request.cookies)

        return {
            // Actual data returned to frontend
            data: {
                msg: 'Hello world!',
                staticImportTestWithExtension,
                staticImportTestWithoutExtension,
                json,
            },
            // Dynamic options for this request
            options: {
                cache: {
                    api: 90,
                },
            },
        }
    },
    // Default static options
    options: {
        cache: {
            api: 85,
        },
    },
}

```

Here are some ideas I have got on community https://github.com/ja-klaudiusz/Nuxt.js-HttpOnly-cookie-auth/blob/master/store/index.js

**Reply by frandiox** [reply]
So as I suspected, there was a bug in development environment where headers were not being forwarded to the props handler during SSR and is fixed in https://github.com/frandiox/vitedge/commit/ebb89045e4dd5721b077d402356f22bd2582d23e

Once this is released, you should be able to access cookies with `headers.get('cookie')`.

To add more info, this is a way to set a cookie once your user is logged in:

```js
  // API endpoint handler
  async handler({ request }) {
    // Passed in the body after the user is logged in (assuming you use some third party JWT provider like Auth0)
    const { jwt } = await request.json()

    // Here you can validate the JWT and extract the expiration date
    const expires = '...'

    return {
      data: {},
      headers: {
        // This should add the expiration date from the actual JWT,
        // and perhaps some domain security: https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies
        'Set-Cookie': `my_auth_key=${jwt}; Path=/; Expires=${expires}; HttpOnly; Secure`,
      },
    }
  }
```

Instead of `my_auth_key` you can call the cookie `bearer` or whatever you prefer. Just use the same cookie name to read it later in the props handler:

```js
  // Props handler
  handler({ headers }) {
    // Parse with 'worktop/cookie' library, for example
    const cookies = parse(headers.get('cookie'))
    
    const jwt = cookies['my_auth_key']
    // Verify JWT, throw Unauthenticated / Forbidden errors if needed
    
    return {
      data: {},
    }
  },
```

**Reply by frandiox** [reply]
@calebeaires Can you try the previous code in `vitedge@0.16.4`? It should work there.

**Reply by calebeaires** [reply]
It works when I set cookies at response but not **getting it back** on request.

I was able to get cookie at request after I set headers manually to the file [node/utils.js](https://github.com/frandiox/vitedge/blob/8e3a493463655d756b5867beae9862b745df1b3b/src/node/utils.js#L50) [line 50]  headers: nodeRequest.headers,


```js
 return new Promise((resolve, reject) => {
        let data = []
        nodeRequest.on('data', (chunk) => data.push(chunk))
        nodeRequest.on('error', (error) => reject(error))
        nodeRequest.on('end', () => {
            resolve(
                new Request(getUrlFromNodeRequest(nodeRequest), {
                    headers: nodeRequest.headers,
                    ...nodeRequest,
                    body: data.length === 0 ? undefined : Buffer.concat(data),
                })
            )
        })
    })


```

Now this works at functions/api:
`console.log(request.headers.get('cookie'))`


Maybe spreed does not pass symbols....!?

**Reply by frandiox** [reply]
@calebeaires Hmmm, it works for me using a local example in this repo 🤔  I've tested reading the cookie in props handlers and also in API handlers without issues.

What version of Node are you using? Also, are you testing with the normal development environment or using something like Express?

I've also checked that `nodeRequest` variable and its `headers` property is enumerable, so it should work normally with the spread operator 🤔

**Reply by calebeaires** [reply]
Node version: v16.8.0
Vitedge: ^0.16.4
Vite: ^2.5.0
Vue: ^3.2.0

Script running: yarn dev (vitedge --ssr --force)

Here is the request I do, now working after setting **manually** the `headers`

![CleanShot 2021-09-20 at 09 03 26@2x](https://user-images.githubusercontent.com/794195/133999047-47ccc920-7b05-4efb-be39-34a9bb5fb524.png)


```ts

// For a JS project, '.js' extension MUST be provided due to ESM limitations
import {ApiEndpoint} from 'vitedge'
import {staticImportTestWithExtension} from '../../utils/test1.js'
// For a TS project, '.js' extension can be optionally provided
import {staticImportTestWithoutExtension} from '../../utils/test2'
// JSON can be imported because --experimental-json-modules is provided in CLI dev
import json from '../../utils/test3.json'

import {JSONCookies} from 'cookie-parser'
import {parse} from "worktop/cookie";


export default <ApiEndpoint>{
    // @ts-ignore
    handler({query, request, headers}) {
        // @ts-ignore

        console.log(' try to get cookie (request.headers) ', request.headers.get('cookie'))
        console.log(' try to get directly (headers) ', headers.get('cookie'))

        if (request.method !== 'GET') {
            throw new Error('Method not supported!')
        }

        return {
            // Actual data returned to frontend
            data: {
                msg: 'Hello world!',
                staticImportTestWithExtension,
                staticImportTestWithoutExtension,
                json,
            },
            // Dynamic options for this request
            options: {
                cache: {
                    api: 90,
                },
            },
        }
    },
    // Default static options
    options: {
        cache: {
            api: 85,
        },
    },
}



```


<img width="998" alt="CleanShot 2021-09-20 at 09 02 11@2x" src="https://user-images.githubusercontent.com/794195/133998867-459b370a-e99b-481a-a4fe-be1ea724b581.png">

But when I set manually headers, the cookie shows there:

![CleanShot 2021-09-20 at 09 04 31@2x](https://user-images.githubusercontent.com/794195/133999203-27fbfb5d-eb64-4b87-ae7b-64fb18093789.png)

**Reply by frandiox** [reply]
@calebeaires Hey sorry, I got quite busy recently because I'm changing jobs. I'll try to come back to this this week.

**Reply by frandiox** [reply]
@calebeaires Hey, can you try `0.17.0`? I hope it works there. There are also more changes to cookies: if you use `fetch` for your API, the cookie header will be automatically injected during SSR.

**Reply by frandiox** [reply]
I've added a guide to the docs: https://vitedge.js.org/recipes/authentication.html

**Reply by calebeaires** [reply]
Hey... @frandiox .... this will change things on my code! so thankful
